File: LeetCode/Java/21.merge-two-sorted-lists.java
/*
 * @lc app=leetcode id=21 lang=java
 *
 * [21] Merge Two Sorted Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode d = new ListNode(-1, l1), tmp;

        while (l1 != null && l2 != null) {
            if (l1.val < l2.val) {
                tmp = l2;
                l2 = l2.next;
                tmp.next = l1.next;
                l1.next = tmp;
                l1 = tmp.next;
            } else {
                tmp = l1;
                l1 = l1.next;
                tmp.next = l2.next;
                l2.next = tmp;
                l2 = tmp.next;
            }
        }
        return d.next;
    }
}
// @lc code=end

File: LeetCode/CPP/21.merge-two-sorted-lists.cpp
// @before-stub-for-debug-begin
#include <vector>
#include <string>
#include "commoncppproblem21.h"

using namespace std;
// @before-stub-for-debug-end

/*
 * @lc app=leetcode id=21 lang=cpp
 *
 * [21] Merge Two Sorted Lists
 */
// struct ListNode
// {
//     int val;
//     ListNode *next;
//     ListNode() : val(0), next(nullptr) {}
//     ListNode(int x) : val(x), next(nullptr) {}
//     ListNode(int x, ListNode *next) : val(x), next(next) {}
// };
// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    // TODO: Fix and Improve
    ListNode *mergeTwoLists(ListNode *l1, ListNode *l2)
    {
        ListNode *d = new ListNode(-1, l1), *tmp = nullptr, *prev = d;
        while (l1 != nullptr && l2 != nullptr)
        {
            if (l1->val < l2->val)
            {
                tmp = l2;
                l2 = l2->next;
                tmp->next = l1->next;
                l1->next = tmp;
                l1 = tmp->next;
                prev = tmp;
            }
            else
            {
                tmp = l2;
                l2 = l2->next;
                tmp->next = l1;
                prev->next = tmp;
                prev = tmp;
                l1 = l1->next;
            }
        }
        ListNode *th = d->next;
        while (th != nullptr)
        {
            cout << th->val << " ";
            th = th->next;
        }
        return d->next;
    }
};
// @lc code=end
